source/nexus.cc

```diff
@@ -89,48 +84,43 @@ G4int main(int argc, char** argv)
   // If there is no other command-line argument to be processed, abort
   // because the user has not provided a configuration macro.
   // (The variable optind is set by getopt_long to the index of the next
-  // element of the argv array to be processed. Once getopt has found all 
-  // of the option arguments, this variable can be used to determine where 
+  // element of the argv array to be processed. Once getopt has found all
+  // of the option arguments, this variable can be used to determine where
   // the remaining non-option arguments begin.)
   if (optind == argc) PrintUsage();
 
   // Assume that the name of the configuration macro is the first
   // command-line parameters that is not a GNU option.
   G4String macro_filename = argv[optind];
 
-  if (macro_filename == "") PrintUsage(); 
-    
-
+  if (macro_filename == "") PrintUsage();
 
   ////////////////////////////////////////////////////////////////////
 
-  
   NexusApp* app = new NexusApp(macro_filename);
   app->Initialize();
 
   G4UImanager* UI = G4UImanager::GetUIpointer();
 
   // if (seed < 0) CLHEP::HepRandom::setTheSeed(time(0));
   // else CLHEP::HepRandom::setTheSeed(seed);
-  
+
   // CLHEP::HepRandom::showEngineStatus();
 
   // visual mode
   if (!batch) {
-    G4VisManager* vismgr = new G4VisExecutive();
-    vismgr->Initialize();
-
-    G4UIsession* session = new G4UIterminal(new G4UItcsh);
+    G4UIExecutive* ui = new G4UIExecutive(argc, argv);
+    G4VisManager* visManager = new G4VisExecutive;
+    visManager->Initialize();
     UI->ApplyCommand("/control/execute macros/vis.mac");
-    session->SessionStart();
-
-    delete session;
-    delete vismgr;
+    ui->SessionStart();
+    delete ui;
+    delete visManager;
```

Comment:
Given that you're touching this code anyway, and that it's of such limited scope that it's impossible not to understand that the change is 100% equivalent ... why not take this opportunity to enter the world of *Modern C++ Programming* (est. 2011) and manage these lifetimes with RAII, specifically `std::unique_ptr`!

# TLDR

``` c++
{
  Banana* my_banana = new Banana(colour, shape, taste);
  // do stuff with my_banana
  delete banana;
}
```
should be written as
``` c++
{
  std::unique_ptr<Banana> my_banana = std::make_unique<Banana>(colour, shape, taste);
  // do stuff with my_banana
} // deleted automatically when unique_ptr goes out of scope
```
or rather
``` c++
{
  auto my_banana = std::make_unique<Banana>(colour, shape, taste);
  // do stuff with my_banana
} // deleted automatically when unique_ptr goes out of scope
```

Caveat: The C++ standards committee forgot to include `make_unique` in C++11, and I believe that Nexus still uses C++11 (that's, like, the middle ages, guys!) So you'll have to write that first line as:

``` c++
std::unique_ptr<Banana> my_banana{colour, shape, taste};
```

Yes, I used braces! Parentheses will work too, in this case, but that's very passé. c.f. [*The Tale of Joe HEP Programmer*](https://github.com/next-exp/nexus/pull/93#issuecomment-809274190).

Lots of goodness follows from using smart pointers, I have proof of this, but there is not enough space in this margin.

# Background

Since C++11 you don't want to see `new` and `delete` *anywhere* in your code, unless you are writing some very low-level utility such as a container, smart pointer or allocator ... which this emphatically is *not*.

Caveats:

+ It's not possible to make `new` disappear completely, but the modern C++ programmer doesn't want to see it unless it is in the argument list of some lifetime-managing consumer.

+ Geant4 makes it *really* hard to make `new`s disappear: it insists you drop all sorts of `new`ed objects on the floor because it takes control of them *implicitly*. Ugh!

+ The C++ committee itself didn't help, by forgetting to include `std::make_unique` until C++14.

Reply:
Thanks for the explanation. But it seems that unique_ptr is not valid in this case or I could be missing something...

```
source/nexus.cc:112:36: error: no matching constructor for initialization of 'std::unique_ptr<G4UIExecutive>'
    std::unique_ptr<G4UIExecutive> ui{argc, argv};
                                   ^ ~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/memory:2500:3: note: candidate constructor template not
      viable: no known conversion from 'int' to 'std::__1::unique_ptr<G4UIExecutive,
      std::__1::default_delete<G4UIExecutive> >::pointer' (aka 'G4UIExecutive *') for 1st argument
  unique_ptr(pointer __p, _LValRefType<_Dummy> __d) _NOEXCEPT
```

Reply:
Ok, I was just following your previous explanation
`std::unique_ptr<Banana> my_banana{colour, shape, taste};`
must be
`std::unique_ptr<Banana> my_banana{new Banana(colour, shape, taste)};`
or maybe, from Joe's tale
`std::unique_ptr<Banana> my_banana{new Banana{colour, shape, taste}};`

Reply:
Hmm, I'm not sure whether you're pointing out an error in my original explanation (if so, I haven't found it yet) or whether you're drawing parallels to my original explanation ... hmm, maybe that's it ... in any case: you are currently stuck in C++11, so let's forget that `std::make_unique` exists (because for you, it won't exist for another 3 years: until 2014).

Let's start with (already using braces for construction wherever possible)

``` c++
G4UIExecutive* ui = new G4UIExecutive{argc, argv};
```

which means

> Create a variable `ui` of type pointer-to-`G4UIExecutive` and initialize it to `new G4UIExecutive{argc, argv}`, and make managing its lifetime be my problem.

First, let's get rid of the annoying and pointless repetition of `G4UIExecutive`; `auto` is your friend:

``` c++
auto ui = new G4UIExecutive{argc, argv};
```

Now we're saying

> Create a variable `ui` and figure out what its type must be, given that we're going to initialize it to `new G4UIExecutive{argc, argv}`, and make managing its lifetime be my problem.

We want to change this to 

> Create a variable `ui` and figure out what its type must be, given that we're going to initialize it to `new G4UIExecutive{argc, argv}`, and let `unique_ptr` manage its lifetime.

The 'let `unique_ptr` manage its lifetime' part is achieved by using it as a construction argument to `unique_ptr`. So we should end up with

``` c++
auto ui = std::unique_ptr<G4UIExecutive>{new std::G4UIExecutive{argc, argv}};
```

Replacing *copy-initialization* with *direct-initialization* we can also write this as

``` c++
std::unique_ptr<G4UIExecutive> ui{new std::G4UIExecutive{argc, argv}};
```

And comparing it to what you wrote in the code that caused the error

``` c++
std::unique_ptr<G4UIExecutive> ui{                       argc, argv };
```

we see that the latter means (very roughly) something like `T* = a, b;` while `T* = new T(a,b)` was needed.

Make sense? Any use to you?

(The nice thing about `make_unique` is that it's a function template, and therefore it can deduce its template argument from its function argument, which spares us the need of repeating `G4UIExecutive`.)

Reply:
Yes, I think it is clear. I was pointing to this, from your first explanation:

Reply:
Hmm ... dunno what that's doing there: it's complete nonsense! Sorry. I'll edit it, lest someone else be misled.